### PR TITLE
gh-108191: Add support of positional argument in SimpleNamespace constructor

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
SimpleNamespace({'a': 1, 'b': 2}) and SimpleNamespace([('a', 1), ('b', 2)]) are now the same as SimpleNamespace(a=1, b=2).

* Issue: gh-108191

----
:books: Documentation preview :books:: https://cpython-previews--108195.org.readthedocs.build/

### CODE REVIEW

File: Objects/namespaceobject.c
Comment (carljm) [inline Objects/namespaceobject.c]: Is it worth centralizing this behavior (check for existence of `keys` method in order to decide whether to merge from mapping or merge from tuple sequence) in a single function? We will now duplicate it both here and in `dict_update_arg`.

**Review by carljm — APPROVED** [review]
This looks like a convenient option to provide in the constructor, with no downside, and established precedent in the dict constructor.

File: Objects/namespaceobject.c
Comment (erlend-aasland) [inline Objects/namespaceobject.c]: ~~"pycore_runtime.h" includes "pycore_global_objects.h", which in turn includes "pycore_global_strings.h":~~

"pycore_global_objects.h" includes "pycore_global_strings.h":

```suggestion
#include "pycore_global_objects.h" // _Py_ID
#include "pycore_runtime.h"       // _Py_ID
```

Comment (erlend-aasland) [reply Objects/namespaceobject.c]: Oh, Victor changed this yesterday in 0dd3fc2a640b273979f94299b545e1e40ac0633c (gh-108228)

Comment (erlend-aasland) [reply Objects/namespaceobject.c]: Oh, well, "pycore_runtime.h" includes "pycore_interp.h" which includes "pycore_global_object.h", so even after Victor's clean-up, you should be fine with only one include:

```suggestion
#include "pycore_runtime.h"       // _Py_ID
```

File: Objects/namespaceobject.c
Comment (serhiy-storchaka) [inline Objects/namespaceobject.c]: It can be simplified to:
```suggestion
            dict = PyObject_CallOneArg((PyObject *)&PyDict_Type, arg);
            if (dict == NULL) {
                return -1;
            }
```

**Review by erlend-aasland — APPROVED** [review]

File: Objects/namespaceobject.c
Comment (serhiy-storchaka) [inline Objects/namespaceobject.c]: With simpler constructor code it will not be needed.
```suggestion
```

File: Objects/namespaceobject.c
Comment (erlend-aasland) [reply Objects/namespaceobject.c]: That is very neat.

File: Objects/namespaceobject.c
Comment (carljm) [reply Objects/namespaceobject.c]: Is it generally preferred to minimize includes like this and rely on transitive includes? I generally thought it was better to directly include anything you need, and not consider the set of includes in another header as part of its public contract.

Comment (erlend-aasland) [reply Objects/namespaceobject.c]: I'm not sure if there is a preferred way of doing includes.

File: Objects/namespaceobject.c
Comment (serhiy-storchaka) [inline Objects/namespaceobject.c]:
```suggestion
        if (!PyArg_ValidateKeywordArguments(dict) ||
```

**Review by erlend-aasland — APPROVED** [review]
Thanks! The "polish" commit was a nice clean-up; it made the code more readable, IMO.

File: Doc/library/types.rst
Comment (AA-Turner) [inline Doc/library/types.rst]:
```suggestion
   Unlike :class:`object`, with :class:`!SimpleNamespace` you can add and remove
```

File: Doc/library/types.rst
Comment (AA-Turner) [inline Doc/library/types.rst]: Should the kwarg behaviour come first, as the longer-established case? I also think initialisation/arguments should be split to a new paragraph, for readability.

File: Lib/test/test_types.py
Comment (AA-Turner) [inline Lib/test/test_types.py]: Is this desirable behaviour? I'd expected either a positional argument or kewword arguments to be legal, but not both at once.

File: Doc/library/types.rst
Comment (serhiy-storchaka) [reply Doc/library/types.rst]: I would be grateful if you simply write new documentation. I understand that my writing is not good, but it is the best I can do.

Comment (AA-Turner) [inline Doc/library/types.rst]: Perhaps:
```suggestion

   :py:class:`SimpleNamespace` objects may be initialized either
   with keyword arguments or with a single positional argument.
   When initialized with keyword arguments,
   those are directly added to the underlying namespace.
   Alternatively, when initialized with a positional argument,
   the underlying namespace will be updated with key-value pairs
   from that argument (either a mapping object or
   an :term:`iterable` object producing key-value pairs).
   All such keys must be strings.
```

Comment (AA-Turner) [reply Doc/library/types.rst]: I've suggested a re-wording!

File: Lib/test/test_types.py
Comment (serhiy-storchaka) [reply Lib/test/test_types.py]: Why not? `dict()` supports this. It is easy to allow it than forbid it.

File: Doc/library/types.rst
Comment (serhiy-storchaka) [reply Doc/library/types.rst]: Thank you!

**Review by AA-Turner — APPROVED** [review]

**Comment by ericsnowcurrently** [comment]
I have no real objections.  I had considered doing this originally but figured it could wait.  Here we are. 😄

**Comment by serhiy-storchaka** [comment]
I tried to implement `SimpleNamespace.__replace__`, and the code may be simpler with this feature:
```python
args = (self.__dict__,)
return type(self)(*args, **kwargs)
```
instead of
```python
d = {}
d.update(self.__dict__)
d.update(kwargs)
return type(self)(**d)
```

**Comment by serhiy-storchaka** [comment]
On other hand, I decided to use other way (#109175), which is more consistent with pickling and copying:
```python
result = type(self)()
result.__dict__.update(self.__dict__)
result.__dict__.update(kwargs)
return result
```
So this is no longer a blocker.

File: Doc/library/types.rst
Comment (AA-Turner) [inline Doc/library/types.rst]:
```suggestion
      Added support for an optional positional argument.
```

or 

```suggestion
      Support an optional positional argument.
```

File: Doc/whatsnew/3.13.rst
Comment (AA-Turner) [inline Doc/whatsnew/3.13.rst]:
```suggestion
* :class:`~types.SimpleNamespace` constructor now allows specifying initial
```

**Review by AA-Turner — COMMENTED** [review]
Two minor grammar points

A

**Review by erlend-aasland — APPROVED** [review]
The updated version looks good to me. I also agree with @AA-Turner's suggestions.

**Comment by serhiy-storchaka** [comment]
Thank you for your suggestions @AA-Turner.

**Review by AA-Turner — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
Seems like this got plenty of approvals, any reason not to merge it?

(I agree with the idea but haven't reviewed the code.)

**Comment by AA-Turner** [comment]
In the issue (https://github.com/python/cpython/issues/108191#issuecomment-1694217768) Serhiy asked for Eric & Raymond's views -- Eric has responded positively but I don't think Raymond has commented either way. 

Though I'd note it's been half a year with no negative comments so I would concur with Jelle.

A

File: Doc/library/types.rst
Comment (ericsnowcurrently) [inline Doc/library/types.rst]:
```suggestion
   :py:class:`SimpleNamespace` objects may be initialized either
   with keyword arguments, with a single positional argument, or with both.
```

Comment (ericsnowcurrently) [reply Doc/library/types.rst]: Perhaps even explicitly point out the similarity to the `dict()` signature, since that may help make thing more clear for readers.

```suggestion
   :py:class:`SimpleNamespace` objects may be initialized
   in the same way as :class:`dict`: either with keyword arguments,
   with a single positional argument, or with both.
```

File: Lib/test/test_types.py
Comment (ericsnowcurrently) [inline Lib/test/test_types.py]: (nothing to change here)

FWIW, I was going to ask why this raises `TypeError`, but I realized it's what `dict()` does in this same situation.

File: Lib/test/test_types.py
Comment (ericsnowcurrently) [inline Lib/test/test_types.py]: Consider adding single-arg-only examples too.

```suggestion
        ns4 = types.SimpleNamespace({'x': 1, 'y': 2}, x=4, z=3)
        ns5 = types.SimpleNamespace([['x', 1], ['y', 2]], x=4, z=3)
        ns6 = types.SimpleNamespace(UserDict({'x': 1, 'y': 2}), x=4, z=3)
        ns7 = types.SimpleNamespace({'x': 1, 'y': 2})
        ns8 = types.SimpleNamespace([['x', 1], ['y', 2]])
```

Maybe also empty-arg examples.

```suggestion
        ns4 = types.SimpleNamespace({'x': 1, 'y': 2}, x=4, z=3)
        ns5 = types.SimpleNamespace([['x', 1], ['y', 2]], x=4, z=3)
        ns6 = types.SimpleNamespace(UserDict({'x': 1, 'y': 2}), x=4, z=3)
        ns7 = types.SimpleNamespace({'x': 1, 'y': 2})
        ns8 = types.SimpleNamespace([['x', 1], ['y', 2]])
        ns9 = types.SimpleNamespace([], x=4, z=3)
        ns10 = types.SimpleNamespace({}, x=4, z=3)
        ns11 = types.SimpleNamespace([])
        ns12 = types.SimpleNamespace({})
```

File: Objects/namespaceobject.c
Comment (ericsnowcurrently) [inline Objects/namespaceobject.c]: Here's another spelling:

```suggestion
        int err = (!PyArg_ValidateKeywordArguments(dict) ||
                   PyDict_Update(ns->ns_dict, dict) < 0);
        Py_DECREF(dict);
        if {err) {
            return -1;
        }
```

...but I'm fine with it the way you have it too.

File: Objects/namespaceobject.c
Comment (ericsnowcurrently) [inline Objects/namespaceobject.c:45]: (nothing to change here)

With this change it seems like `namespace_init()` is now almost identical `dict.__init__()` and `dict.update()`.  Could we take advantage of that?  Probably not since we have to verify string keys.

FWIW, I was going to make some suggestions (incl. exposing `dict_update_common()` in the internal C-API), until I remembered the string keys thing.  Maybe that function could have an option to check for string keys first (or a variant function that checks)?  However, it probably wouldn't be worth it without other use cases beside `SimpleNamespace`.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]

File: Doc/whatsnew/3.13.rst
Comment (eendebakpt) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  an iterable of key-value pairs.
```

**Comment by serhiy-storchaka** [comment]
I have made the requested changes; please review again.

File: Doc/library/types.rst
Comment (ericsnowcurrently) [inline Doc/library/types.rst:505]: It may be outside the scope of this PR, but I noticed that in gh-109175 you didn't add `__replace__()` to the "equivalent" code here.  It's probably fine that this isn't an exact equivalence, especially since we say "roughly equivalent".

I just wanted to mention it (before I forget), in case it's something that would make sense to address in this PR.  I expect it would basically be what you wrote in https://github.com/python/cpython/issues/109174#issuecomment-1712430549.

Either way, I'll leave it up to you.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

I left one comment about something mostly unrelated to this PR, so I don't consider it a blocker.

File: Doc/library/types.rst
Comment (serhiy-storchaka) [reply Doc/library/types.rst:505]: It also does not have `__reduce__()`. Both methods would be simple, but I am not sure that it is needed to add them in this example. They may shadow the main idea of the class.

**Review by erlend-aasland — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
It would be good to get @rhettinger's feedback, but this PR already received many positive feedbacks, and I already invested in this issue so much, that I'm going to merge it anyway, even if initially I was not so interested in this.

**Review by carljm — APPROVED** [review]
Seems to be a reasonable amount of interest in this, and plenty of review approvals. Can we go ahead and merge for 3.13?

**Comment by carljm** [comment]
Looks like a merge conflict needs to be resolved.

**Pull request merged**
